### PR TITLE
Add schema-serde support for awsQuery and awsQuery-compatible

### PR DESCRIPTION
**Pull request by vcjana**
## Motivation and Context

Adds schema-based serialization/deserialization for the `awsQuery` and `awsQuery-compatible` protocols, building on the schema-serde infrastructure from #4660.

## Description

**awsQuery-compatible** (`aws-smithy-json`):
- Wraps `AwsJsonRpcProtocol`, adds `x-amzn-query-mode: true` request header
- Error code extraction from `x-amzn-query-error` response header (codegen)

**awsQuery** (`aws-smithy-query` + `aws-smithy-xml`):
- `QueryShapeSerializer`: URL-encoded form body serialization with dot-notation nesting, `@xmlName`, `@xmlFlattened` support
- `XmlShapeDeserializer`: XML response deserialization with struct/list/map support, `@xmlName` resolution, `is_null()` returns false (XML has no null concept)
- `AwsQueryProtocol`: implements `ClientProtocolInner` directly (asymmetric format — form-encoded request, XML response)
- Strips `<OpResponse><OpResult>` envelope in `deserialize_response`

**Codegen wiring**:
- `SchemaDecorator`: protocol matching for both `AwsQueryTrait` and `AwsQueryCompatibleTrait`
- `ResponseDeserializerGenerator`: `x-amzn-query-error` header parsing
- `ClientProtocolTestGenerator` / `SerdeBenchmarkTestGenerator`: awsQuery branches

**Verification**:
- Cross-validated serializer output byte-for-byte against existing `QueryWriter`
- Benchmarked at parity (~600ns/iter, <1% overhead vs `QueryWriter`)
- Enabled all protocols in allowlist locally — no regressions in restJson1, awsJson1.0, awsJson1.1

## Testing

- 36 tests in `aws-smithy-query` (unit, cross-validation, edge cases, protocol)
- 37 tests in `aws-smithy-xml` (deserializer: struct, nested struct, list, map, xmlName, is_null, scalars)
- 3 tests in `aws-smithy-json` (awsQuery-compatible protocol)
- Generated awsQuery code compiles clean with allowlist enabled

### CODE REVIEW

File: rust-runtime/aws-smithy-xml/src/codec/deserializer.rs [hunk 1]
```diff
@@ -0,0 +1,483 @@
...
+fn read_text(scope: &mut ScopedDecoder<'_, '_>) -> Result<String, SerdeError> {
+    try_data(scope)
+        .map(|cow| cow.into_owned())
+        .map_err(|e| SerdeError::InvalidInput {
+            message: e.to_string(),
+        })
+}
+
+impl<'a> ShapeDeserializer for XmlShapeDeserializer<'a> {
+    fn read_struct(
```
Comment (landonxjames) [inline rust-runtime/aws-smithy-xml/src/codec/deserializer.rs:179]: Couple of broad issues with struct deserialization here:

First it seems like nested structs with more than one level of struct nesting are losing data. The existing `read_nested_struct` test gets around this by only having a single struct with nested simple values. So this test fails:

```rust
    #[test]
    fn nested_struct_three_levels_deep() {
        static LEAF: Schema =
            Schema::new_member(shape_id!("t", "Inner"), ShapeType::String, "Leaf", 0);
        static INNER_SCHEMA: Schema =
            Schema::new_struct(shape_id!("t", "Inner"), ShapeType::Structure, &[&LEAF]);
        static INNER_MEMBER: Schema =
            Schema::new_member(shape_id!("t", "Middle"), ShapeType::Structure, "Inner", 0);
        static MIDDLE_SCHEMA: Schema = Schema::new_struct(
            shape_id!("t", "Middle"),
            ShapeType::Structure,
            &[&INNER_MEMBER],
        );
        static MIDDLE_MEMBER: Schema =
            Schema::new_member(shape_id!("t", "Outer"), ShapeType::Structure, "Middle", 0);
        static OUTER_SCHEMA: Schema = Schema::new_struct(
            shape_id!("t", "Outer"),
            ShapeType::Structure,
            &[&MIDDLE_MEMBER],
        );
        static OUTER_MEMBER: Schema =
            Schema::new_member(shape_id!("t", "Root"), ShapeType::Structure, "Outer", 0);
        static ROOT: Schema = Schema::new_struct(
            shape_id!("t", "Root"),
            ShapeType::Structure,
            &[&OUTER_MEMBER],
        );

        let xml = b"<Root><Outer><Middle><Inner><Leaf>value</Leaf></Inner></Middle></Outer></Root>";
        let mut deser = XmlShapeDeserializer::new(xml);
        let mut leaf = String::new();
        deser
            .read_struct(&ROOT, &mut |outer_m, d_outer| {
                assert_eq!(outer_m.member_name(), Some("Outer"));
                d_outer.read_struct(&OUTER_SCHEMA, &mut |middle_m, d_middle| {
                    assert_eq!(middle_m.member_name(), Some("Middle"));
                    d_middle.read_struct(&MIDDLE_SCHEMA, &mut |inner_m, d_inner| {
                        assert_eq!(inner_m.member_name(), Some("Inner"));
                        d_inner.read_struct(&INNER_SCHEMA, &mut |leaf_m, d_leaf| {
                            if leaf_m.member_name() == Some("Leaf") {
                                leaf = d_leaf.read_string(leaf_m)?;
                            }
                            Ok(())
                        })
                    })
                })
            })
            .expect("3-level nested struct should round-trip");
        assert_eq!(leaf, "value");
    }
```

There are similar issues with `read_list` and `read_map`. Pretty sure the cause is more or less the same for all of them. They reconstruct sub-trees by reading the text of immediate children with `read_text` (which calls `try_data`) and re-emitting that text inside fabricated tags via `format!("<{}>{}</{}>", name, text, name)`. This pattern has two issues:
* errors out as soon as a child has element children of its own, because `try_data` returns Err on a non-text token
* unescapes leaf text before splicing it into reconstructed markup, so any payload containing `&` or `<` produces invalid XML on re-parse.

You can see that second behavior in this test:

```rust
#[test]
    fn struct_member_with_escaped_text_round_trips() {
        static VALUE: Schema =
            Schema::new_member(shape_id!("t", "Body"), ShapeType::String, "value", 0);
        static BODY_SCHEMA: Schema =
            Schema::new_struct(shape_id!("t", "Body"), ShapeType::Structure, &[&VALUE]);
        static PAYLOAD_MEMBER: Schema =
            Schema::new_member(shape_id!("t", "Envelope"), ShapeType::Structure, "payload", 0);
        static ENVELOPE_SCHEMA: Schema = Schema::new_struct(
            shape_id!("t", "Envelope"),
            ShapeType::Structure,
            &[&PAYLOAD_MEMBER],
        );

        // Server response with an XML-escaped `&` in the leaf value. After
        // `unescape`, the original value is `foo&bar`.
        let xml = b"<Envelope><payload><value>foo&amp;bar</value></payload></Envelope>";
        let mut deser = XmlShapeDeserializer::new(xml);
        let mut value_str = String::new();
        deser
            .read_struct(&ENVELOPE_SCHEMA, &mut |payload_m, d_payload| {
                let _ = payload_m;
                d_payload.read_struct(&BODY_SCHEMA, &mut |inner_m, d_inner| {
                    if inner_m.member_name() == Some("value") {
                        value_str = d_inner.read_string(inner_m)?;
                    }
                    Ok(())
                })
            })
            .expect("escaped text inside a nested struct should round-trip");
        assert_eq!(value_str, "foo&bar");
    }
```

File: rust-runtime/aws-smithy-xml/src/codec/deserializer.rs [hunk 2]
```diff
@@ -0,0 +1,483 @@
...
+
+fn read_text(scope: &mut ScopedDecoder<'_, '_>) -> Result<String, SerdeError> {
+    try_data(scope)
+        .map(|cow| cow.into_owned())
+        .map_err(|e| SerdeError::InvalidInput {
+            message: e.to_string(),
+        })
+}
+
+impl<'a> ShapeDeserializer for XmlShapeDeserializer<'a> {
```
Comment (landonxjames) [inline rust-runtime/aws-smithy-xml/src/codec/deserializer.rs:178]: Deserializer should have a max recursion depth, see [CBOR](https://github.com/smithy-lang/smithy-rs/pull/4667/changes#diff-fa0e07df72330bc08285f97c4c5613cf13d119319991c1863919b12eaa0a4f33R31) and [JSON](https://github.com/smithy-lang/smithy-rs/blob/adc0a46bad5a34fb77e088377eb294c907e013d8/rust-runtime/aws-smithy-json/src/codec/deserializer.rs#L21) deserializers for an example of what that could look like.

File: rust-runtime/aws-smithy-query/src/protocol.rs [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+        &mut self,
+        schema: &Schema,
+        consumer: &mut dyn FnMut(String, &mut dyn ShapeDeserializer) -> Result<(), SerdeError>,
+    ) -> Result<(), SerdeError> {
+        let mut deser = aws_smithy_xml::codec::XmlShapeDeserializer::new(self.data.as_bytes());
+        deser.read_map(schema, consumer)
+    }
+
+    fn read_boolean(&mut self, schema: &Schema) -> Result<bool, SerdeError> {
+        aws_smithy_xml::codec::XmlShapeDeserializer::new(self.data.as_bytes()).read_boolean(schema)
```
Comment (landonxjames) [inline rust-runtime/aws-smithy-query/src/protocol.rs]: Should we be creating a new `XmlShapeDeserializer` for every single type? Couldn't `OwnedXmlDeserializer` just own the inner deserializer?

File: rust-runtime/aws-smithy-query/src/protocol.rs [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+        inner.push_str(&format!("<{}>{}</{}>", name, text, name));
+    }
+    Ok(format!("<R>{}</R>", inner))
+}
+
+struct OwnedXmlDeserializer {
+    data: String,
+}
+
+impl ShapeDeserializer for OwnedXmlDeserializer {
```
Comment (landonxjames) [inline rust-runtime/aws-smithy-query/src/protocol.rs]: Not entirely convinced we need this type at all, why isn't `XmlShapeDeserializer` sufficient?

File: rust-runtime/aws-smithy-xml/src/codec/deserializer.rs [hunk 5]
```diff
@@ -0,0 +1,483 @@
...
+            while let Some(mut inner) = entry_tag.next_tag() {
+                let local = inner.start_el().local().to_string();
+                let text = read_text(&mut inner)?;
+                match local.as_str() {
+                    "key" => key = text,
+                    "value" => value_text = text,
+                    _ => {}
+                }
+            }
+            if !key.is_empty() {
```
Comment (landonxjames) [inline rust-runtime/aws-smithy-xml/src/codec/deserializer.rs]: I'm fairly certain that the empty string `""` is a valid map key so this would drop those items.

File: rust-runtime/aws-smithy-query/src/protocol.rs [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+    }
+}
+
+fn strip_aws_query_envelope(xml: &str) -> Result<String, SerdeError> {
+    use aws_smithy_xml::decode::Document as XmlDoc;
+    let mut doc = XmlDoc::new(xml);
+    let mut response_el = doc.root_element().map_err(|e| SerdeError::InvalidInput {
+        message: format!("invalid XML response: {e}"),
+    })?;
+    // First child is <OpResult>
```
Comment (landonxjames) [inline rust-runtime/aws-smithy-query/src/protocol.rs]: Is this guaranteed? I'm having trouble finding anything in the spec that asserts this will always be first. This check should probably be name-based rather than positional.

File: rust-runtime/aws-smithy-xml/src/codec/deserializer.rs [hunk 7]
```diff
@@ -0,0 +1,483 @@
...
+fn resolve_xml_member<'a>(schema: &'a Schema, element_name: &str) -> Option<&'a Schema> {
+    schema
+        .members()
+        .iter()
+        .find(|m| {
+            let wire_name = m
+                .xml_name()
+                .map(|t| t.value())
+                .or(m.member_name())
+                .unwrap_or("");
```
Comment (landonxjames) [inline rust-runtime/aws-smithy-xml/src/codec/deserializer.rs]: This feels like it should probably be an error rather than silently inserting the empty string?

File: codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/generators/protocol/ResponseDeserializerGenerator.kt [hunk 8]
```diff
@@ -475,6 +477,15 @@ class ResponseDeserializerGenerator(
                 "BoxError" to RuntimeType.boxError(runtimeConfig),
                 "error_symbol" to errorSymbol,
             )
+            if (codegenContext.serviceShape.hasTrait<AwsQueryCompatibleTrait>()) {
+                rustTemplate(
+                    """
+                    let error_code = headers.get("x-amzn-query-error")
```
Comment (landonxjames) [inline codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/generators/protocol/ResponseDeserializerGenerator.kt]: I'm not 100% sure this is correct, at least probably warrants a comment pointing to the [spec](https://smithy.io/2.0/aws/protocols/aws-query-protocol.html#awsqueryerror-trait). It feels like we shouldn't just be throwing away half of the `x-amzn-query-error` header

File: codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/customizations/SchemaDecorator.kt
Comment (landonxjames) [inline codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/customizations/SchemaDecorator.kt]: You will need to add your new protocols to the `allowedProtocols` list to get them to actually generate code and run integration/protocol tests against them. Will likely encounter a bunch of CI failures that will help you debug when enabling that.

Will want to see at least one successful CI run of the final code before a final commit to remove the changes from the allowlist and then merge

File: rust-runtime/aws-smithy-query/src/protocol.rs [hunk 9]
```diff
@@ -0,0 +1,329 @@
...
+
+        let uri = if endpoint.is_empty() { "/" } else { endpoint };
+        let mut request = Request::new(SdkBody::from(body));
+        request
+            .set_method("POST")
+            .map_err(|e| SerdeError::custom(format!("{e}")))?;
+        request
+            .set_uri(uri)
+            .map_err(|e| SerdeError::custom(format!("{e}")))?;
+        request
```
Comment (landonxjames) [inline rust-runtime/aws-smithy-query/src/protocol.rs:65]: This method mostly seems to be re-implementing the logic from `aws-smithy-schema/src/schema/http_protocol/rpc.rs`. Is there a way we could make the existing logic more reusable?

It is also dropping the `Content-Length` header. Is that intentional? I see it at various places in the spec [ex](https://smithy.io/2.0/aws/protocols/aws-query-protocol.html#operation-error-serialization)

File: rust-runtime/aws-smithy-query/src/codec/serializer.rs [hunk 10]
```diff
@@ -0,0 +1,710 @@
...
+    fn write_integer(&mut self, schema: &Schema, value: i32) -> Result<(), SerdeError> {
+        self.write_scalar(schema, &value.to_string())
+    }
+
+    fn write_long(&mut self, schema: &Schema, value: i64) -> Result<(), SerdeError> {
+        self.write_scalar(schema, &value.to_string())
+    }
+
+    fn write_float(&mut self, schema: &Schema, value: f32) -> Result<(), SerdeError> {
+        self.write_scalar(schema, &value.to_string())
```
Comment (landonxjames) [inline rust-runtime/aws-smithy-query/src/codec/serializer.rs]: Pretty sure `f32::to_string()` outputs `-inf/inf` for infinity, but the [spec](https://smithy.io/2.0/aws/protocols/aws-query-protocol.html#non-numeric-float-and-double-serialization) says the serialized values should be `-Infinity/Infinity` (applies to `f64` below as well)

File: rust-runtime/aws-smithy-query/src/codec/serializer.rs [hunk 11]
```diff
@@ -0,0 +1,710 @@
...
+            self.push_prefix(self.wire_name(schema));
+        }
+        value.serialize_members(self)?;
+        if is_member {
+            self.pop_prefix();
+        }
+        Ok(())
+    }
+
+    fn write_list(
```
Comment (landonxjames) [inline rust-runtime/aws-smithy-query/src/codec/serializer.rs:168]: Pretty sure this doesn't work for lists of structs since they don't emit an element index?

```rust
#[test]
    fn list_of_struct_emits_per_element_index() {
        static NAME: Schema =
            Schema::new_member(shape_id!("test", "Item"), ShapeType::String, "Name", 0);
        static ITEM_SCHEMA: Schema =
            Schema::new_struct(shape_id!("test", "Item"), ShapeType::Structure, &[&NAME]);
        static ITEMS_LIST: Schema =
            Schema::new_member(shape_id!("test", "I"), ShapeType::List, "Items", 0);

        struct Item(&'static str);
        impl SerializableStruct for Item {
            fn serialize_members(&self, s: &mut dyn ShapeSerializer) -> Result<(), SerdeError> {
                s.write_string(&NAME, self.0)
            }
        }

        let mut ser = QueryShapeSerializer::new("Op", "1.0");
        ser.write_list(&ITEMS_LIST, &|s| {
            s.write_struct(&ITEM_SCHEMA, &Item("X1"))?;
            s.write_struct(&ITEM_SCHEMA, &Item("X2"))
        })
        .unwrap();

        // awsQuery wire format spec says each list element is qualified
        // with its 1-based index between the list name and the inner field.
        assert_eq!(
            String::from_utf8(ser.finish()).unwrap(),
            "Action=Op&Version=1.0\
             &Items.member.1.Name=X1\
             &Items.member.2.Name=X2",
        );
    }
```

Similarish bug would happen for list of lists

```rust
#[test]
    fn list_of_list_uses_per_outer_index_and_no_empty_segments() {
        static OUTER: Schema =
            Schema::new_member(shape_id!("test", "I"), ShapeType::List, "Outer", 0);
        static INNER_LIST: Schema =
            Schema::new_list(shape_id!("test", "InnerList"), &aws_smithy_schema::prelude::STRING);

        let mut ser = QueryShapeSerializer::new("Op", "1.0");
        ser.write_list(&OUTER, &|s_outer| {
            s_outer.write_list(&INNER_LIST, &|s_inner| {
                s_inner.write_string(&aws_smithy_schema::prelude::STRING, "a")?;
                s_inner.write_string(&aws_smithy_schema::prelude::STRING, "b")
            })?;
            s_outer.write_list(&INNER_LIST, &|s_inner| {
                s_inner.write_string(&aws_smithy_schema::prelude::STRING, "c")
            })
        })
        .unwrap();

        // Per the awsQuery spec, `List<List<X>>` named `Outer` serializes as
        // `Outer.member.<i>.member.<j>=v` — the outer list's element index
        // sits between the two `.member` segments.
        assert_eq!(
            String::from_utf8(ser.finish()).unwrap(),
            "Action=Op&Version=1.0\
             &Outer.member.1.member.1=a\
             &Outer.member.1.member.2=b\
             &Outer.member.2.member.1=c",
        );
    }
```

File: rust-runtime/aws-smithy-query/src/codec/serializer.rs [hunk 12]
```diff
@@ -0,0 +1,710 @@
...
+        write_elements(self)?;
+        self.context_stack.pop();
+        if !flat {
+            self.pop_prefix();
+        }
+        self.pop_prefix();
+        Ok(())
+    }
+
+    fn write_map(
```
Comment (landonxjames) [inline rust-runtime/aws-smithy-query/src/codec/serializer.rs:232]: Think there is a similar bug to the list thing for maps, but don't quite have time to write a test right now.

**Review by landonxjames — COMMENTED** [review]
Couple bugs in the serde that need looking at. Would suggest enabling your protocols in the allowlist and seeing what fails in CI to get a start, suspect the protocol tests would catch most of this.

File: codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/customizations/SchemaDecorator.kt
Comment (landonxjames) [reply codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/customizations/SchemaDecorator.kt]: Follow up based on @ysaito1001 comment the other day, should probably enable all supported protocols to ensure that none of the codegen changes broke anything for the other protocols

File: rust-runtime/aws-smithy-xml/src/codec/deserializer.rs (hunk 1 above)
Comment (vcjana) [reply rust-runtime/aws-smithy-xml/src/codec/deserializer.rs:179]: Rewrote the deserializer tp use `xmlparser` byte offsets now instead of `try_data + XML` reconstruction which handles arbitrary nesting and preserves escaping. Both tests pass.

File: rust-runtime/aws-smithy-xml/src/codec/deserializer.rs (hunk 2 above)
Comment (vcjana) [reply rust-runtime/aws-smithy-xml/src/codec/deserializer.rs:178]: added `XmlDeserializerSettings` with configurable `max_depth`

File: rust-runtime/aws-smithy-query/src/protocol.rs (hunk 3 above)
Comment (vcjana) [reply rust-runtime/aws-smithy-query/src/protocol.rs]: Yeah looked into this the deserializer borrows from the owned String, so storing both in the same struct would be self-referential. Creating a new one per call avoids that.

File: rust-runtime/aws-smithy-query/src/protocol.rs (hunk 4 above)
Comment (vcjana) [reply rust-runtime/aws-smithy-query/src/protocol.rs]: same as [above](https://github.com/smithy-lang/smithy-rs/pull/4668#discussion_r3276708632) self-referential lifetime issue. The stripped envelope is a new `String` that needs the owner

File: rust-runtime/aws-smithy-xml/src/codec/deserializer.rs (hunk 5 above)
Comment (vcjana) [reply rust-runtime/aws-smithy-xml/src/codec/deserializer.rs]: good catch removed the guard. empty keys are accepted

File: rust-runtime/aws-smithy-query/src/protocol.rs (hunk 6 above)
Comment (vcjana) [reply rust-runtime/aws-smithy-query/src/protocol.rs]: changed to find the element by name `*Result` or `Error` suffix instead of assuming position.

File: rust-runtime/aws-smithy-xml/src/codec/deserializer.rs (hunk 7 above)
Comment (vcjana) [reply rust-runtime/aws-smithy-xml/src/codec/deserializer.rs]: changed to return `None` (member won't match) instead of falling back to empty string

File: codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/generators/protocol/ResponseDeserializerGenerator.kt (hunk 8 above)
Comment (vcjana) [reply codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/generators/protocol/ResponseDeserializerGenerator.kt]: added a comment. Only the code portion is needed for dispatch same as [parse_aws_query_compatible_error](https://github.com/smithy-lang/smithy-rs/blob/main/rust-runtime/inlineable/src/aws_query_compatible_errors.rs#L18).

File: rust-runtime/aws-smithy-query/src/protocol.rs (hunk 9 above)
Comment (vcjana) [reply rust-runtime/aws-smithy-query/src/protocol.rs:65]: Added Content-Length. Can't reuse `HttpRpcProtocol` awsQuery needs `Metadata` from `cfg` for the `Action=` param, and the response format (XML) differs from the request (form-encoded)

File: rust-runtime/aws-smithy-query/src/codec/serializer.rs (hunk 10 above)
Comment (vcjana) [reply rust-runtime/aws-smithy-query/src/codec/serializer.rs]: fixed matches on special values per spec now

File: rust-runtime/aws-smithy-query/src/codec/serializer.rs (hunk 11 above)
Comment (vcjana) [reply rust-runtime/aws-smithy-query/src/codec/serializer.rs:168]: Fixed in [e48aae2d0](https://github.com/smithy-lang/smithy-rs/commit/e48aae2d0)

File: rust-runtime/aws-smithy-query/src/codec/serializer.rs (hunk 12 above)
Comment (vcjana) [reply rust-runtime/aws-smithy-query/src/codec/serializer.rs:232]: fixed in [e48aae2d0]( https://github.com/smithy-lang/smithy-rs/pull/4668/changes/e48aae2d0bd73c082023c08b9b9304e834a433f8#diff-ae26369cbbeb04f6ff2daa82c8ec6d5cfc817a501160ec1ce19ce74d7705731cR870)

File: codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/generators/protocol/ResponseDeserializerGenerator.kt [hunk 13]
```diff
@@ -475,6 +477,17 @@ class ResponseDeserializerGenerator(
                 "BoxError" to RuntimeType.boxError(runtimeConfig),
                 "error_symbol" to errorSymbol,
             )
+            if (codegenContext.serviceShape.hasTrait<AwsQueryCompatibleTrait>()) {
```
Comment (ysaito1001) [inline codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/generators/protocol/ResponseDeserializerGenerator.kt]: This part feels a bit procedural where the logic is driven by if-else branching on the protocol rather than leveraging polymorphism. Ideally, since this is a protocol-agnostic code path, we'd like to avoid "if I'm this protocol, do X" style logic here. As it stands, any future protocol that requires custom error handling would mean adding yet another if-else branch.

Also, lines 483–487 appear to duplicate [existing logic](https://github.com/smithy-lang/smithy-rs/blob/24622dae1baa50f5b98eba4c0cfabfcc828237b7/rust-runtime/inlineable/src/aws_query_compatible_errors.rs#L11-L23) — could we somehow reuse that instead?

File: rust-runtime/aws-smithy-json/src/protocol/aws_query_compatible.rs [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+        &self,
+        input: &dyn SerializableStruct,
+        input_schema: &Schema,
+        endpoint: &str,
+        cfg: &ConfigBag,
+    ) -> Result<Request, SerdeError> {
+        let mut request = self
+            .inner
+            .serialize_request(input, input_schema, endpoint, cfg)?;
+        request.headers_mut().insert("x-amzn-query-mode", "true");
```
Comment (ysaito1001) [inline rust-runtime/aws-smithy-json/src/protocol/aws_query_compatible.rs]: A reverse question of [this](https://github.com/smithy-lang/smithy-rs/pull/4667#discussion_r3257522943). Can we get away without `.headers_mut().insert(...)` ?

File: rust-runtime/aws-smithy-query/src/protocol.rs (hunk 4 above)
Comment (ysaito1001) [reply rust-runtime/aws-smithy-query/src/protocol.rs]: If `strip_aws_query_envelope` returned byte offsets into the original response body instead of building a new string, could we borrow directly from response and use `XmlShapeDeserializer<'a>` without introducing `OwnedXmlDeserializer`?

Even if it doesn't pan out, it seems worth exploring as a way to avoid introducing `OwnedXmlDeserializer`.

**Review by ysaito1001 — COMMENTED** [review]
Good start! A couple of clarifying questions.

File: codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/customizations/SchemaDecorator.kt [hunk 15]
```diff
@@ -48,7 +51,14 @@ object SchemaSerdeAllowlist {
...
      */
-    private val allowedProtocols: Set<ShapeId> = emptySet()
+    private val allowedProtocols: Set<ShapeId> =
+        setOf(
+            RestJson1Trait.ID,
+            AwsJson1_0Trait.ID,
+            AwsJson1_1Trait.ID,
+            AwsQueryTrait.ID,
+            AwsQueryCompatibleTrait.ID,
+        )
```
Comment (ysaito1001) [inline codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/customizations/SchemaDecorator.kt]: Just leaving a reminder to empty this set before merging this PR to avoid turning on schema serde in main.

File: codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/generators/protocol/ResponseDeserializerGenerator.kt (hunk 13 above)
Comment (vcjana) [reply codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/generators/protocol/ResponseDeserializerGenerator.kt]: added `resolve_error_code` to `ClientProtocolInner`, codegen calls it polymorphically now

File: rust-runtime/aws-smithy-json/src/protocol/aws_query_compatible.rs (hunk 14 above)
Comment (vcjana) [reply rust-runtime/aws-smithy-json/src/protocol/aws_query_compatible.rs]: moved to `static_headers()` on ClientProtocolInner. protocol declares its headers, `serialize_request` applies them from the trait.

File: codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/customizations/SchemaDecorator.kt (hunk 15 above)
Comment (vcjana) [reply codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/customizations/SchemaDecorator.kt]: addressed in [31ac810](https://github.com/smithy-lang/smithy-rs/pull/4668/changes/31ac810a318adfa1d0b0f253a80b6105529e6c18)

File: rust-runtime/aws-smithy-query/src/protocol.rs (hunk 4 above)
Comment (vcjana) [reply rust-runtime/aws-smithy-query/src/protocol.rs]: it panned out `strip_aws_query_envelope` returns &str offsets now, borrows directly from response. OwnedXmlDeserializer is gone.

**Comment by ysaito1001** [comment]
My first take after revisiting the latest version of this PR:
 
1. Split into two PRs: one for awsQuery, one for awsQuery-compatible. They should be largely independent (?) and helps reduce the amount of code to review in each PR.
2. Rebase the awsQuery PR on top of #4686: I think the XML deserialization side can benefit from XmlDeserializer in that PR.
